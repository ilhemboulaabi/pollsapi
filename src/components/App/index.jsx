import {
  BrowserRouter as Router,
  Route,
  Switch,
} from 'react-router-dom';
import React, { Suspense, lazy } from 'react';
import { LinearProgress } from '@material-ui/core';

function App() {
  const Main = lazy(() => import('../Main'));
  const QuestionDetails = lazy(() => import('../QuestionDetails'));
  const addQuestion = lazy(() => import('../AddQuestion'));
  return (
    <Router>
      <Suspense fallback={<LinearProgress />}>
        <Switch>
          <Route exact path="/" component={Main} />
          <Route path="/questions/:questionId" component={QuestionDetails} />
          <Route path="/add" component={addQuestion} />
        </Switch>
      </Suspense>
    </Router>
  );
}

export default App;
