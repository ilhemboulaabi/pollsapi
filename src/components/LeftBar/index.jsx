import React from 'react';
import {
  makeStyles,
  Button,
} from '@material-ui/core';
import AddIcon from '@material-ui/icons/Add';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import Select from '../Select';

const useStyles = makeStyles((theme) => ({
  container: {
    display: 'flex',
    flexDirection: 'column',
    marginRight: theme.spacing(2),
  },
  addButton: {
    marginBottom: theme.spacing(4),
  },
  link: {
    decoration: 'none',
    color: theme.palette.grey[800],
  },
}));

const LeftBar = ({ handelSort }) => {
  const classes = useStyles();
  const options = [
    {
      key: 'default',
      value: 'published_at',
      text: 'Select...',
    },
    {
      key: 'question',
      value: 'question',
      text: 'Question',
    },
  ];

  return (
    <div className={classes.container}>
      <Button
        variant="contained"
        color="default"
        className={classes.addButton}
        startIcon={<AddIcon />}
      >
        <Link to="/add" className={classes.link}>Add Question</Link>
      </Button>
      <Select
        id="sort"
        label="Sort Questions"
        action={handelSort}
        options={options}
      />
    </div>
  );
};

export default LeftBar;

LeftBar.propTypes = {
  handelSort: PropTypes.func,
};
LeftBar.defaultProps = {
  handelSort: () => {},
};
