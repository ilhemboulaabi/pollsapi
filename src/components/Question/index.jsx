import React from 'react';
import { makeStyles } from '@material-ui/core/styles';
import {
  Card,
  CardMedia,
  CardContent,
  Typography,
  Grid,
  CardActions,
  Button,
  LinearProgress,
} from '@material-ui/core';
import PropTypes from 'prop-types';
import Moment from 'moment';
import { Link } from 'react-router-dom';

const useStyles = makeStyles(() => ({
  card: {
    display: 'flex',
  },
  container: {
    display: 'flex',
    flexDirection: 'column',
    width: '60%',
  },
  media: {
    width: 140,
  },
}));

const Question = ({ question, loading }) => {
  const classes = useStyles();

  return (
    loading
      ? <LinearProgress />
      : (
        <Grid item xs={12}>
          <Card className={classes.card} raised>
            <CardMedia
              className={classes.media}
              image="https://cdn.pixabay.com/photo/2015/11/03/08/56/question-mark-1019820_960_720.jpg"
              title="Contemplative Reptile"
            />
            <div className={classes.container}>
              <CardContent>
                <Typography component="h5" variant="h5">
                  {question.question}
                </Typography>
                <Typography variant="subtitle1" color="textSecondary">
                  {Moment(question.published_at).format('d MMM yyyy')}
                </Typography>
              </CardContent>
              <CardActions>
                <Button size="small" color="primary">
                  <Link to={`${question.url}`}>Vote here!</Link>
                </Button>
              </CardActions>
            </div>
          </Card>
        </Grid>
      )
  );
};

export default Question;

Question.propTypes = {
  // eslint-disable-next-line react/forbid-prop-types
  question: PropTypes.object,
  loading: PropTypes.bool,
};
Question.defaultProps = {
  question: {},
  loading: false,
};
