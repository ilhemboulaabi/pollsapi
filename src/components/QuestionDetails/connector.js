import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import * as QuestionDetailsActions from '../../store/actions/questionDetails';

const mapStateToProps = (state) => ({
  questionDetails: state.questionDetails.questionDetails,
  isQuestionDetailsLoading: state.questionDetails.isQuestionDetailsLoading,
});

const mapDispatchToProps = (dispatch) => ({
  actions: {
    questionDetails: bindActionCreators(QuestionDetailsActions, dispatch),
  },
});

export default connect(mapStateToProps, mapDispatchToProps);
