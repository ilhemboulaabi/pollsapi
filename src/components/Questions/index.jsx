import React from 'react';
import {
  Grid,
  LinearProgress,
  Box,
} from '@material-ui/core';
import Alert from '@material-ui/lab/Alert';
import PropTypes from 'prop-types';
import Question from '../Question';

const Questions = ({ questions, loading }) => (
  loading || (questions === undefined || questions === null || questions.length === 0)
    ? (
      <Box>
        <LinearProgress />
        <Alert severity="info">No Data available!</Alert>
      </Box>
    )
    : (
      <Grid container spacing={2}>
        {
        questions && questions.map((question) => (
          <Question
            key={question.url}
            question={question}
          />
        ))
        }
      </Grid>
    )
);

export default Questions;

Questions.propTypes = {
  // eslint-disable-next-line react/forbid-prop-types
  questions: PropTypes.array,
  loading: PropTypes.bool,
};
Questions.defaultProps = {
  questions: [],
  loading: false,
};
