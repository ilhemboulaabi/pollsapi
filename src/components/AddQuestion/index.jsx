import React, { Component } from 'react';
import { withStyles } from '@material-ui/core/styles';
import {
  TextField,
  Button,
  Typography,
  Tooltip,
} from '@material-ui/core';
import PropTypes from 'prop-types';
import Alert from '@material-ui/lab/Alert';
import connector from './connector';

const useStyles = (theme) => ({
  root: {
    display: 'flex',
    flexDirection: 'column',
    margin: theme.spacing(2),
    '& > *': {
      margin: theme.spacing(1),
      width: '25ch',
    },
  },
});

export class AddQuestion extends Component {
  constructor(props) {
    super(props);
    this.state = {
      question: '',
      answers: '',
      error: false,
    };
  }

  handleQuestionChange = (event) => {
    this.setState({
      question: event.target.value,
    });
  }

  handleAnswerChange = (event) => {
    this.setState({
      answers: event.target.value,
    });
  }

  handleSubmit = () => {
    const { question, answers } = this.state;
    const answersArray = answers.split('\n');
    const { history, actions } = this.props;
    actions.questions.addQuestion(question, answersArray).then(() => {
      this.setState({
        question: '',
        answers: '',
      });
      history.goBack();
    }, () => {
      this.setState({
        error: true,
      });
    });
  };

  render() {
    const { classes } = this.props;
    const { error } = this.state;

    return (
      <>
        {error && <Alert severity="warning">Check your data out! Check ToolTip!</Alert>}
        <form className={classes.root} noValidate autoComplete="off">
          <Typography component="h5" variant="h5">
            Add a question
          </Typography>
          <TextField
            required
            id="standard-basic"
            label="Question!"
            onChange={this.handleQuestionChange}
          />
          <Tooltip title="Type minimum two answers, Type 'Enter' after each answer">
            <TextField
              required
              id="outlined-multiline-static"
              label="Answers!"
              multiline
              rows={4}
              variant="outlined"
              onChange={this.handleAnswerChange}
            />
          </Tooltip>
          <Button variant="contained" onClick={this.handleSubmit}>Submit</Button>
        </form>
      </>
    );
  }
}

export default withStyles(useStyles, { withTheme: true })(connector(AddQuestion));

AddQuestion.propTypes = {
  // eslint-disable-next-line react/forbid-prop-types
  actions: PropTypes.object,
  // eslint-disable-next-line react/forbid-prop-types
  classes: PropTypes.object,
  // eslint-disable-next-line react/forbid-prop-types
  history: PropTypes.object,
};
AddQuestion.defaultProps = {
  actions: {},
  classes: {},
  history: {},
};
