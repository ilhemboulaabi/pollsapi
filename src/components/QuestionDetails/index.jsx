import React, { Component } from 'react';
import {
  List,
  ListItem,
  ListItemText,
  ListItemAvatar,
  Avatar,
  LinearProgress,
  Box,
  Typography,
  IconButton,
} from '@material-ui/core';
import Alert from '@material-ui/lab/Alert';
import ThumbUpIcon from '@material-ui/icons/ThumbUp';
import PropTypes from 'prop-types';
import connector from './connector';

export class QuestionDetails extends Component {
  componentDidMount() {
    const { match, actions } = this.props;
    actions.questionDetails.loadQuestionDetails(match.url);
  }

  render() {
    const { questionDetails, isQuestionDetailsLoading } = this.props;

    return (
      isQuestionDetailsLoading
        ? (
          <Box>
            <LinearProgress />
            <Alert severity="info">No Data available!</Alert>
          </Box>
        )
        : (
          <Box m={2}>
            <Typography component="h5" variant="h5">
              {questionDetails.question}
            </Typography>
            <List>
              {
                questionDetails.choices && questionDetails.choices.map((choice) => (
                  <ListItem key={choice.url}>
                    <ListItemAvatar>
                      <Avatar>
                        <IconButton color="primary" aria-label="upload picture" component="span">
                          <ThumbUpIcon />
                        </IconButton>
                      </Avatar>
                    </ListItemAvatar>
                    <ListItemText primary={choice.choice} secondary={`${choice.votes} vote`} />
                  </ListItem>
                ))
              }
            </List>
          </Box>
        )
    );
  }
}

export default connector(QuestionDetails);

QuestionDetails.propTypes = {
  // eslint-disable-next-line react/forbid-prop-types
  questionDetails: PropTypes.object,
  isQuestionDetailsLoading: PropTypes.bool,
  // eslint-disable-next-line react/forbid-prop-types
  match: PropTypes.object,
  // eslint-disable-next-line react/forbid-prop-types
  actions: PropTypes.object,
};
QuestionDetails.defaultProps = {
  questionDetails: {},
  isQuestionDetailsLoading: false,
  match: {},
  actions: {},
};
