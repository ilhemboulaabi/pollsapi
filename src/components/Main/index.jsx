import React, { Component } from 'react';
import { withStyles } from '@material-ui/core/styles';
import {
  Grid,
} from '@material-ui/core';
import Pagination from '@material-ui/lab/Pagination';
import PropTypes from 'prop-types';
import TopBar from '../TopBar';
import LeftBar from '../LeftBar';
import Questions from '../Questions';
import connector from './connector';

const useStyles = (theme) => ({
  root: {
    padding: theme.spacing(2),
    [theme.breakpoints.up('md')]: {
      padding: theme.spacing(2, 20, 2, 20),
    },
  },
  pagination: {
    marginTop: theme.spacing(2),
  },
});

export class Main extends Component {
  constructor(props) {
    super(props);
    this.state = { currentPage: 1 };
  }

  componentDidMount() {
    const { actions } = this.props;
    const { currentPage } = this.state;
    actions.questions.loadQuestions(currentPage);
  }

  componentDidUpdate(prevProps, prevState) {
    const { actions } = this.props;
    const { currentPage } = this.state;
    if (prevState.currentPage !== currentPage) {
      actions.questions.loadQuestions(currentPage);
    }
  }

  handelPageChange = (event, value) => {
    const { currentPage } = this.state;
    if (currentPage !== value) {
      this.setState({
        currentPage: value,
      });
    }
  }

  render() {
    const {
      questions, isQuestionsLoading, actions, classes,
    } = this.props;
    const { currentPage } = this.state;

    return (
      <>
        <TopBar
          handelSearch={actions.questions.searchQuestion}
        />
        <Grid className={classes.root} container>
          <Grid item xs={2} sm={3} md={4}>
            <LeftBar
              handelSort={actions.questions.sortQuestions}
            />
          </Grid>
          <Grid item xs={10} sm={9} md={8}>
            <Questions
              questions={questions}
              loading={isQuestionsLoading}
            />
            <Pagination
              className={classes.pagination}
              count={2}
              page={currentPage}
              onChange={this.handelPageChange}
            />
          </Grid>
        </Grid>
      </>
    );
  }
}

export default withStyles(useStyles, { withTheme: true })(connector(Main));

Main.propTypes = {
  // eslint-disable-next-line react/forbid-prop-types
  questions: PropTypes.array,
  isQuestionsLoading: PropTypes.bool,
  // eslint-disable-next-line react/forbid-prop-types
  actions: PropTypes.object,
  // eslint-disable-next-line react/forbid-prop-types
  classes: PropTypes.object,
};
Main.defaultProps = {
  questions: [],
  isQuestionsLoading: false,
  actions: {},
  classes: {},
};
