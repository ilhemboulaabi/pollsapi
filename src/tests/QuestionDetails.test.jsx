import * as React from 'react';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { shallowToJson } from 'enzyme-to-json';
import { QuestionDetails } from '../components/QuestionDetails';

Enzyme.configure({ adapter: new Adapter() });

const setup = () => {
  const props = {
    QuestionDetails: {},
    isQuestionDetailsLoading: false,
    actions: {
      QuestionDetails: {
        loadQuestionDetails: jest.fn(),
      },
    },
    match: {
      url: '/questions/1',
    },
  };

  const wrapper = shallow(
    <QuestionDetails {...props} />,
    {
      disableLifecycleMethods: true,
    },
  );

  return {
    props,
    wrapper,
  };
};

describe('QuestionDetails', () => {
  it('should render self and subcomponents', () => {
    const { wrapper } = setup();
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });
});
