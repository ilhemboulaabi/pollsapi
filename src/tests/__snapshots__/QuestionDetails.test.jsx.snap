// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QuestionDetails should render self and subcomponents 1`] = `
<Styled(MuiBox)
  m={2}
>
  <WithStyles(ForwardRef(Typography))
    component="h5"
    variant="h5"
  />
  <WithStyles(ForwardRef(List)) />
</Styled(MuiBox)>
`;
